esmvaltool/cmorizers/data/cmorizer.py

```diff
@@ -0,0 +1,536 @@
+        start : str, optional
+            Start of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        end : str, optional
+            End of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        overwrite : bool, optional
+            If true, download already present data again
+        """
+        start = self._parse_date(start)
+        end = self._parse_date(end)
+
+        self.formatter.start('download', datasets, config_file, kwargs)
+        self.formatter.download(start, end, overwrite)
+
+    def format(self,
+               datasets,
+               config_file=None,
+               start=None,
+               end=None,
+               install=False,
+               **kwargs):
+        """Format datasets.
+
+        Parameters
+        ----------
+        datasets : list(str)
+            List of datasets to format
+        config_file : str, optional
+            Path to ESMValTool's config user file, by default None
+        start : str, optional
+            Start of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        end : str, optional
+            End of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        install : bool, optional
+            If true, move processed data to the folder, by default False
+        """
+        start = self._parse_date(start)
+        end = self._parse_date(end)
+
+        self.formatter.start('formatting', datasets, config_file, kwargs)
+        self.formatter.format(start, end, install)
+
+    def prepare(self,
+                datasets,
+                config_file=None,
+                start=None,
+                end=None,
+                overwrite=False,
+                install=False,
+                **kwargs):
+        """Download and format a set of datasets.
+
+        Parameters
+        ----------
+        datasets : list(str)
+            List of datasets to format
+        config_file : str, optional
+            Path to ESMValTool's config user file, by default None
+        start : str, optional
+            Start of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        end : str, optional
+            End of the interval to process, by default None. Valid formats
+            are YYYY, YYYYMM and YYYYMMDD.
+        install : bool, optional
+            If true, move processed data to the folder, by default False
+        overwrite : bool, optional
+            If true, download already present data again
+        """
+        start = self._parse_date(start)
+        end = self._parse_date(end)
+
+        self.formatter.start('preparation', datasets, config_file, kwargs)
+        if not self.formatter.download(start, end, overwrite):
+            return False
```

Comment:
What's the purpose of this return? Ok, something went wrong with the download. But didn't that raise an informative error/exception? If not, are we in principal good to go? As far as I could understand, the effect of this `return False` is just that "False" will be printed (and that the command terminates normally). That doesn't strike me as particularly helpful. But for symmetry then, we might add a `return True` at the end of the function?

Reply:
I think the point is to break the prepare (download + format) command: if the dataset cannot be downloaded for some reason then the format (cmorization) function is not called. I think the current exceptions in the Downloader class are not sufficient for that. But I'll have a deeper look tomorrow to see how we could improve that.

Reply:
I had a deeper look and I think we can get rid of the `return False` (and not add `return True`). Both returns would output unnecessary `True` and `False` in the logs.

The question is: if a user includes a non-supported file in the list of datasets for the `prepare` command (download + format), e.g. `esmvaltool data prepare TEST,WOA,PHC`, do we want the tool to:

1.  Download all supported datasets (PHC, WOA) and stop without formatting any data?
2. Download all supported datasets and then format all supported datasets? In this second case, this will output two types of error:
```bash
    raise ValueError(
ValueError: Dataset TEST does not have an automatic downloader

    raise Exception(
Exception: Format failed for datasets TEST
```
Note that the second error is identical to the one obtained when formatting a non-supported dataset `esmvaltool data format TEST`.

I don't have a strong opinion on this. Perhaps a first error is a good reason not to format data. But if a user wants to use `prepare` then they expect data formatting right after data downloading.

Code for both cases:
1. Data downloading of all supported datasets and stop:
```python
if self.formatter.download(start, end, overwrite):
    self.formatter.format(start, end, install)
```
2. Data downloading and formatting of all supported datasets
```python
self.formatter.download(start, end, overwrite)
self.formatter.format(start, end, install)
```

Reply:
I gave it a second thought and opted for option 1, meaning that the formatting will be done only if the list of datasets passed as an argument of `esmvaltool data prepare` is successfully downloaded. I could imagine cases where the user does `esmvaltool data prepare MODIS` for a dataset that only has a formatter script but no downloader script. With option 2, it could be that the tool finds raw data (depending on the config-user file) and rewrites previously cmorized data without the user noticing. I think only datasets for which we have downloader and a formatter scripts should be handled successfully with `esmvaltool data prepare`.